vendor/users/models/user.server.model.js

```diff
@@ -430,7 +431,47 @@ UserSchema.statics.generateRandomPassphrase = function generateRandomPassphrase(
   });
 };
 
+/**
+ * Creates a TTL index to remove users having expired/non validated emails.
+ * @see https://docs.mongodb.com/manual/core/index-ttl/
+ */
+
+const createEmailExpirationIndex = async (userModel) => {
+  const { email: emailConfig } = config.validations.config;
+
+  if (emailConfig.validate && emailConfig.expiration_timeout > 0) {
+
+    const filterExpression = {
+      'validations.type': 'email',
+      'validations.validated': false,
+    };
+
+    // Remove outdated index
+    const indexes = await userModel.collection.indexes();
+    const outdatedIndex = indexes.find(({ name, partialFilterExpression, expireAfterSeconds }) => 
+      name.startsWith('created_at')
+      && partialFilterExpression
+      && _.isEqual(partialFilterExpression, filterExpression)
+      && expireAfterSeconds !== emailConfig.expiration_timeout,
+    );
+
+    if (outdatedIndex) {
+      userModel.collection.dropIndex(outdatedIndex.name);
+    }
+
+    // Create index (if it doesn't exist)
+    userModel.collection.createIndex(
+      { created_at: 1 },
+      {
+        expireAfterSeconds: emailConfig.expiration_timeout,
+        partialFilterExpression: filterExpression,
+      },
+    );
+  }
+};
+
 const UserModel = mongoose.model('User', UserSchema);
 UserModel.createIndexes();
+createEmailExpirationIndex(UserModel);
```

Comment:
I thought about making this function more generic, by creating a utility function which take care of checking the existence of the index & override it if needed, but i'm worried about making the code harder to read this way (especially the outdated index matching conditions, which has to be passed to the generic function as a parameter) :thinking: